h/api/search/config.py

```diff
@@ -0,0 +1,334 @@
+    _update_index_analysis(client.conn, index, analysis)
+    _update_index_mappings(client.conn, index, mappings)
+
+
+def _ensure_icu_plugin(conn):
+    """Ensure that the ICU analysis plugin is installed for ES."""
+    # Pylint issue #258: https://bitbucket.org/logilab/pylint/issue/258
+    #
+    # pylint: disable=unexpected-keyword-arg
+    names = [x.strip() for x in conn.cat.plugins(h='component').split('\n')]
+    if 'analysis-icu' not in names:
+        message = ("ICU Analysis plugin is not installed for Elasticsearch\n"
+                   "  See the installation instructions for more details:\n"
+                   "  https://github.com/hypothesis/h/blob/master/"
+                   "INSTALL.rst#installing")
+        raise RuntimeError(message)
+
+
+def _append_config(mappings, analysis, doc_type, type_mappings, type_analysis):
+    """
+    Append config for the named type to pre-existing config.
+
+    This function takes a mappings dict and an analysis dict which may contain
+    prior data, and attempts to update them with mappings and analysis settings
+    for the named type.
+    """
+    mappings.update({doc_type: type_mappings})
+    for section, items in type_analysis.items():
+        existing_items = analysis.setdefault(section, {})
+        for name in items:
+            if name in existing_items:
+                fmt = "Duplicate definition of 'index.analysis.{}.{}'."
+                msg = fmt.format(section, name)
+                raise RuntimeError(msg)
+        existing_items.update(items)
+
+
+def _create_index(conn, name, settings):
+    """
+    Create index with the specific name and settings.
+
+    This function will ignore errors caused by the index already existing.
+    """
+    # Check if the index exists (perhaps as an alias) and if so, return.
+    if conn.indices.exists(index=name):
+        return
+
+    # Otherwise, try to create the index
+    conn.indices.create(name, body=settings)
+
+
+def _update_index_analysis(conn, name, analysis):
+    """Attempt to update the index analysis settings."""
+    name = _resolve_alias(conn, name)
+    settings = conn.indices.get_settings(index=name)
+    existing = settings[name]['settings']['index'].get('analysis', {})
+    if existing != analysis:
+        try:
+            conn.indices.close(index=name)
+            conn.indices.put_settings(index=name, body={
+                'analysis': analysis
+            })
+        finally:
+            conn.indices.open(index=name)
+
+
+def _update_index_mappings(conn, name, mappings):
+    """Attempt to update the index mappings."""
+    name = _resolve_alias(conn, name)
+    try:
+        for doc_type, body in mappings.items():
+            conn.indices.put_mapping(index=name,
+                                     doc_type=doc_type,
+                                     body=body)
+    except elasticsearch.exceptions.RequestError as e:
+        if not e.error.startswith('MergeMappingException'):
```

Comment:
Nope, this is the only way of checking it. These errors simply wrap the semi-structured error data which comes back from ElasticSearch.